### PR TITLE
Plotting + Map updates

### LINKED ISSUE #76: Task: Event Type Pie Chart
This output will be a set of EcoPlot pie-charts  (one for every grouper combination) showing the proportion (and count) of the different event types

### LINKED ISSUE #77: Task: Stacked Bar Chart
This output will be a single EcoPlot stacked bar chart of event types over time (the x-axis grouping will be set as a param  e.g., by day, week, month etc)

### LINKED ISSUE #78: Task: Create EcoMap with multiple GDFs (e.g., Tracks & Events)
This task may be trivial after the merged dataframes. Need to check with @atmorling

### PR DESCRIPTION
**Pull request by atmorling**
Closes #76
Closes #77 
Closes #78

Raising as a draft for review - merging is blocked until the below PRs in `ecoscope` are merged
https://github.com/wildlife-dynamics/ecoscope/pull/230
https://github.com/wildlife-dynamics/ecoscope/pull/237

- Adds tasks for creating stacked bar chart and pie chart
- Adds `create_map_layer` tasks and updates the `draw_ecomap` interface to suit

### CODE REVIEW

**Comment by cisaacstern** [comment]
@atmorling are unit tests failing because of https://github.com/wildlife-dynamics/ecoscope/pull/237 presumably?

(Lint is broken, as identified in #124, I will try to fix that today.)

**Comment by atmorling** [comment]
Yup, the ecoscope dep needs a bump once the two aforementioned PRs are in.

**Review by cisaacstern — COMMENTED** [review]
One small comment below.

Didn't want to let perfect be the enemy of the good, so throwing this one out there first, and will get to a fuller review shortly.

**Comment by cisaacstern** [comment]
Lint workflow issue is fixed by #129.

P.S., sorry for the merge conflicts from #117. They should hopefully not be too bad to resolve by:
- ❌ Deleting `examples/compilation-specs/time-density.yaml`
- ❌ Deleting `examples/dags/time_density_dag.jupytext.py`
- ❌ Deleting `examples/dags/time_density_dag.script_sequential.py`
- ❌ Deleting `examples/params/time_density_params.yaml`
- ❌ Deleting `examples/params/time_density_params_fillable.json`
- ❌ Deleting `examples/params/time_density_params_fillable.yaml`
- ♻️ Recompiling and fixing conflicts in  `examples/compilation-specs/patrol_workflow.yaml`

**Comment by atmorling** [comment]
> P.S., sorry for the merge conflicts from #117. They should hopefully not be too bad to resolve by:

Serves me right for not taking a closer look at that PR :joy:

File: ecoscope_workflows/tasks/results/_ecoplot.py [hunk 1]
```diff
@@ -59,3 +59,142 @@ def draw_ecoplot(
...
+        Field(
+            description="Style arguments passed to plotly.graph_objects.Bar and applied to individual groups."
+        ),
+    ] = None,
+    style_kws: Annotated[
+        dict,
+        Field(
+            description="Style arguments passed to plotly.graph_objects.Bar and applied to all groups."
+        ),
+    ] = {},
```
[the remark this answers is not in the document]
Comment (atmorling) [reply ecoscope_workflows/tasks/results/_ecoplot.py]: I've updated this pr to remove the use of `{}` as a default in plot/map tasks

File: examples/compilation-specs/patrol_workflow.yaml [hunk 2]
```diff
@@ -40,11 +46,17 @@ workflow:
     task: apply_reloc_coord_filter
     with:
       df: ${{ workflow.patrol_events.return }}
+  - name: Create map layer from Patrols Events
+    id: patrol_events_map_layer
+    task: create_map_layer
+    with:
+      geodataframe: ${{ workflow.filter_patrol_events.return }}
   - name: Draw Ecomap for Patrols Events
```
[the remark this answers is not in the document]
Comment (atmorling) [reply examples/compilation-specs/patrol_workflow.yaml:56]: Will do ASAP!

File: examples/compilation-specs/patrol_workflow.yaml [hunk 3]
```diff
@@ -42,11 +48,17 @@ workflow:
     task: apply_reloc_coord_filter
     with:
       df: ${{ workflow.patrol_events.return }}
+  - name: Create map layer from Patrols Events
```
[the remark this answers is not in the document]
Comment (atmorling) [reply examples/compilation-specs/patrol_workflow.yaml:51]: So an example of how that would look is:
```yaml
- name: Create map layer from Patrols Events
  id: patrol_events_map_layer
  task: create_map_layer
  with:
    geodataframe: ${{ workflow.filter_patrol_events.return }}
- name: Create map layer from Time Density
  id: td_map_layer
  task: create_map_layer
  with:
    geodataframe: ${{ workflow.td.return }}
- name: Draw Ecomap from Time Density
  id: td_ecomap
  task: draw_ecomap
  with:
    geo_layers:
      - ${{ workflow.patrol_events_map_layer.return }}
      - ${{ workflow.td_map_layer.return }}
```

[the remark this answers is not in the document]
Comment (atmorling) [reply examples/compilation-specs/patrol_workflow.yaml:51]: It's probably worth adding test coverage for that scenario - I'll take a look at doing that

[the remark this answers is not in the document]
Comment (Yun-Wu) [reply examples/compilation-specs/patrol_workflow.yaml:51]: Yea, my secret goal is for the `patrol_workflow.yaml` to cover MVP after all the tasks are closed.

[the remark this answers is not in the document]
Comment (cisaacstern) [reply examples/compilation-specs/patrol_workflow.yaml:51]: I **_love_** this secret goal 💯

**Review by cisaacstern — COMMENTED** [review]
LGTM, @atmorling!

I enthusiastically second @Yun-Wu's suggestions re: getting multilayers + pie + stacked bar in the patrols spec. Otherwise, good to go from my perspective! TYSM!

File: examples/compilation-specs/patrol_workflow.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (atmorling) [reply examples/compilation-specs/patrol_workflow.yaml]: Done!

**Comment by atmorling** [comment]
So working through the review comments today I found a couple problems with `stacked_bar_chart`:

1. The task needs an extra parameter to handle the time interval (ie; month, day, hour, etc) to 'bin' results into. This logic should exist here and not in `ecoscope`, but it is made _significantly_ cleaner/simpler once this [this](https://github.com/wildlife-dynamics/ecoscope/pull/231) `ecoscope` PR is merged. 

We can either hold off on this PR and address it once `ecoscope` has been updated, or raise a separate issue and address it in a future PR. I'm in favour of the latter just to unblock more things, but happy either way.

In either case I think it's also worth making the task more specific/explicit by calling it `time_series_bar_chart` and renaming some params (like `x_axis`) param `time_column` or similar. This might be yak shaving though so I'm keen to hear what others think. 

2. (less severe) There's a [bug](https://github.com/wildlife-dynamics/ecoscope/issues/242). This isn't  a blocker, since we want to provide style  parameters in the patrols dashboard context, but worth fixing ASAP.

**Comment by cisaacstern** [comment]
> We can either hold off on this PR and address it once ecoscope has been updated, or raise a separate issue and address it in a future PR. I'm in favour of the latter just to unblock more things, but happy either way.

> In either case I think it's also worth making the task more specific/explicit by calling it time_series_bar_chart and renaming some params (like x_axis) param time_column or similar. This might be yak shaving though so I'm keen to hear what others think.

@atmorling on both counts, I would say if there's a way to merge this by Thursday that doesn't result in a lot of extra/unnecessary tech debt, let's aim for that.

Ideally on Friday we have all tasks in place to run e2e patrols workflow.

**Comment by cisaacstern** [comment]
> merge this by Thursday

@atmorling to be more specific, if at all possible, could we aim for merge by Friday morning US Pacific Time? My goal is that when I start work Friday I can converge the concurrent threads of tasks + #128 to finish out an e2e patrols run.

But if that timeline means tech debt that would be a lot easier to avoid in the first place, then let's discuss.

**Comment by atmorling** [comment]
Sorry I didn't mean to imply that any of this was to be 'kicked down the road', I should have said 'in an adjacent PR' rather than 'future'.
I'll have the changes in tomorrow and if no one has any issues with the `ecoscope` [pr](https://github.com/wildlife-dynamics/ecoscope/pull/231) by then I'll merge that too.

**Comment by cisaacstern** [comment]
Ok cool! No worries I may have misread due to being a bit distracted by #128. Excited to see this go in!

**Review by cisaacstern — APPROVED** [review]
Amazing! 🤩 LGTM

**Comment by cisaacstern** [comment]
(Oh looks like there are just a few conflicts to resolve following #123 )

**Pull request merged**